File: backend/internal/handlers/web/cpuIssue.go
package web

import (
	"encoding/json"
	"log"

	. "openreplay/backend/pkg/messages"
	"openreplay/backend/pkg/messages/performance"
)

/*
	Handler name: CpuIssue
	Input events: PerformanceTrack,
				  SetPageLocation
	Output event: IssueEvent
*/

const CPU_THRESHOLD = 70 // % out of 100
const CPU_MIN_DURATION_TRIGGER = 6 * 1000

type CpuIssueDetector struct {
	startTimestamp uint64
	startMessageID uint64
	lastTimestamp  uint64
	maxRate        uint64
	contextString  string
}

func (f *CpuIssueDetector) Build() Message {
	if f.startTimestamp == 0 {
		return nil
	}
	duration := f.lastTimestamp - f.startTimestamp
	timestamp := f.startTimestamp
	messageID := f.startMessageID
	maxRate := f.maxRate

	f.startTimestamp = 0
	f.startMessageID = 0
	f.maxRate = 0
	if duration < CPU_MIN_DURATION_TRIGGER {
		return nil
	}

	payload, err := json.Marshal(struct {
		Duration uint64
		Rate     uint64
	}{duration, maxRate})
	if err != nil {
		log.Printf("can't marshal CpuIssue payload to json: %s", err)
	}

	return &IssueEvent{
		Type:          "cpu",
		Timestamp:     timestamp,
		MessageID:     messageID,
		ContextString: f.contextString,
		Payload:       string(payload),
	}
}

func (f *CpuIssueDetector) Handle(message Message, messageID uint64, timestamp uint64) Message {
	switch msg := message.(type) {
	case *PerformanceTrack:
		dt := performance.TimeDiff(timestamp, f.lastTimestamp)
		if dt == 0 {
			return nil // TODO: handle error
		}

		f.lastTimestamp = timestamp

		if msg.Frames == -1 || msg.Ticks == -1 {
			return f.Build()
		}

		cpuRate := performance.CPURate(msg.Ticks, dt)

		if cpuRate >= CPU_THRESHOLD {
			if f.startTimestamp == 0 {
				f.startTimestamp = timestamp
				f.startMessageID = messageID
			}
			if f.maxRate < cpuRate {
				f.maxRate = cpuRate
			}
		} else {
			return f.Build()
		}
	case *SetPageLocation:
		f.contextString = msg.URL
	}
	return nil
}

File: tracker/tracker/src/webworker/detectors/cpuIssue.ts
import Message, {
  Type,
  PerformanceTrack,
} from '../../common/messages.gen.js'
import type { Detector, ReportIssue } from './types.js'
import { cpuRate, timeDiff } from './performance.js'

/**
 * Port of backend/pkg/handlers/web/cpuIssue.go.
 *
 * Tracks stretches where CPU load (derived from PerformanceTrack ticks) stays
 * above CPU_THRESHOLD for at least CPU_MIN_DURATION_TRIGGER ms.
 *
 * NOTE: the Go version overwrites lastTimestamp *before* computing the interval,
 * so its dt is always 0 (cpuRate always 0 -> effectively disabled). We compute
 * dt against the previous sample so the rate is meaningful.
 */

const CPU_THRESHOLD = 70 // % out of 100
const CPU_MIN_DURATION_TRIGGER = 6 * 1000

export default class CpuIssueDetector implements Detector {
  readonly types: readonly number[] = [Type.PerformanceTrack]

  private startTimestamp = 0
  private startMessageId = 0
  private lastTimestamp = 0
  private maxRate = 0

  constructor(
    private readonly report: ReportIssue,
    private readonly url: () => string,
  ) {}

  private duration(): number {
    return this.lastTimestamp - this.startTimestamp
  }

  private reset(): void {
    this.startTimestamp = 0
    this.startMessageId = 0
    this.maxRate = 0
  }

  private build(): void {
    const start = this.startTimestamp
    const messageId = this.startMessageId
    const rate = this.maxRate
    const duration = this.duration()
    // matches Go `defer f.reset()`
    this.reset()
    if (start === 0 || duration < CPU_MIN_DURATION_TRIGGER) {
      return
    }
    this.report({
      type: 'cpu',
      contextString: this.url(),
      payload: JSON.stringify({ Duration: duration, Rate: rate }),
      timestamp: start,
      messageId,
    })
  }

  flush(_timestamp: number): void {
    this.build()
  }

  handle(message: Message, index: number, timestamp: number): void {
    const msg = message as PerformanceTrack
    const frames = msg[1]
    const ticks = msg[2]
    if (timestamp < this.lastTimestamp) {
      return
    }
    const prevTimestamp = this.lastTimestamp
    this.lastTimestamp = timestamp
    if (prevTimestamp === 0) {
      // first sample: no interval to measure the rate against yet
      return
    }
    const rate = cpuRate(ticks, timeDiff(timestamp, prevTimestamp))
    if (frames === -1 || ticks === -1 || rate < CPU_THRESHOLD) {
      this.build()
      return
    }
    if (this.startTimestamp === 0) {
      this.startTimestamp = timestamp
      this.startMessageId = index
    }
    if (this.maxRate < rate) {
      this.maxRate = rate
    }
  }
}
